### PR TITLE
mgr/dashboard: unselect rows in datatables

### PR DESCRIPTION
**Pull request by Sarthak0702**
User should be able to unselect a selected row in data-tables.
referred from: https://github.com/swimlane/ngx-datatable/issues/775#issuecomment-304007717

Fixes: https://tracker.ceph.com/issues/53244, https://tracker.ceph.com/issues/54575
Signed-off-by: Sarthak0702 <sarthak.0702@gmail.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by nizamial09 — APPROVED** [review]
LGTM! Thanks @Sarthak0702

**Comment by nizamial09** [comment]
Is it possible to verify this in a small unit test? @Sarthak0702

**Comment by Sarthak0702** [comment]
> Is it possible to verify this in a small unit test? @Sarthak0702

Thats a good idea, will try to add that.

**Comment by Sarthak0702** [comment]
jenkins test windows

**Comment by Sarthak0702** [comment]
jenkins test make check

**Comment by Sarthak0702** [comment]
jenkins test make check

**Comment by Sarthak0702** [comment]
> This fix breaks the multi select functionality in data-tables (Cluster-> OSDs ). Need to investigate this.

Fixed. As unselect functionality is  natively supported by ngx-datatables for `'multiclick' & 'multi'`, `selectCheck` is only added for `'single' selectionType`.

**Comment by Sarthak0702** [comment]
jenkins test dashboard

**Comment by Sarthak0702** [comment]
jenkins test dashboard cephadm

**Comment by votdev** [comment]
Nice idea, but i would like to see that this feature behaves the way a desktop user is used to from operating systems like Windows or Gnome, especially Windows Explorer, Nautilus or other file navigation tools. There you can unselect a row by using STRG + LMB.

**Review by votdev — DISMISSED** [review]
Nice idea, but i would like to see that this feature behaves the way a desktop user is used to from operating systems like Windows or Gnome, especially Windows Explorer, Nautilus or other file navigation tools. There you can unselect a row by using CTRL + LMB. It is not good to introduce a different workflow instead of implementing a widely used and accepted standard.

**Comment by votdev** [comment]
> > Nice idea, but i would like to see that this feature behaves the way a desktop user is used to from operating systems like Windows or Gnome, especially Windows Explorer, Nautilus or other file navigation tools. There you can unselect a row by using CTRL + LMB. It is not good to introduce a different workflow instead of implementing a widely used and accepted standard.
> 
> So are you suggesting to use `LMB` to select a row but `CTRL+LMB` to unselect? Wouldn't that be a bit confusing for the user?

Yes, that is the standard in Windows and Linux UIs.

**Comment by Sarthak0702** [comment]
> Yes, that is the standard in Windows and Linux UIs.

Hmm, I see. FYI, for `selectionType:'multiClick'` (check Cluster -> OSDs) in ngx-datable (on which TableComponent is built) `LMB` is used to select and unselect rows natively. So we need to do disable that somehow too.

**Comment by votdev** [comment]
> > Yes, that is the standard in Windows and Linux UIs.
> 
> Hmm, I see. FYI, for `selectionType:'multiClick'` (check Cluster -> OSDs) in ngx-datable (on which TableComponent is built) `LMB` is used to select and unselect rows natively. So we need to do disable that somehow too.

Yes i know. It's not my decision which style the Dashboard follows, but i think it should go the same way as the most common operating systems and file navigators go. I mean, the LMB approach to select and unselect will work on touch devices, but i don't think someone is using the Dashboard on those devices.

**Comment by Sarthak0702** [comment]
> Yes i know, this was IMO also not implemented the right way. But it's not my decision which style the Dashboard follows, but i think it should go the same way as the most common operating systems and file navigators do. I mean, the LMB approach to select and unselect will work on touch devices, but i don't think someone is using the Dashboard on those devices.

Honesty I do find using CTRL+ LMB to unselect rows a bit inorganic in a web based application. Have a look at this: https://www.patternfly.org/v4/components/table#composable-row-click-handler-hoverable--selected-rows
Also, surprisingly there are some folks who do use dashboard on touch devices (like iPads).  Nevertheless, I and @epuertat would agree, feel that the dashboard tables look a bit outdated, and some modern UI changes would surely improve UX. For this I have opened a tracker issue: https://tracker.ceph.com/issues/55095

**Review by epuertat — APPROVED** [review]
LGTM! Thanks @Sarthak0702 !

File: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts [hunk 1]
```diff
@@ -707,6 +711,10 @@ export class TableComponent implements AfterContentChecked, OnInit, OnChanges, O
     this.updateSelection.emit(_.clone(this.selection));
   }
 
+  private singleSelectCheck(row: any) {
+    return this.selection.selected.indexOf(row) === -1;
+  }
```
[the remark this answers is not in the document]
Comment (Sarthak0702) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts:716]: Ahh, I see. This is a good point, will keep this in mind.

**Review by nizamial09 — APPROVED** [review]

**Comment by Sarthak0702** [comment]
jenkins test dashboard

**Comment by Sarthak0702** [comment]
jenkins test dashboard cephadm

**Comment by Sarthak0702** [comment]
jenkins test dashboard cephadm

**Review by nizamial09 — CHANGES_REQUESTED** [review]
@Sarthak0702 the dashboard cephadm e2e is failing for real. The below solution will fix it.

```
--- a/src/pybind/mgr/dashboard/frontend/cypress/integration/cluster/hosts.po.ts
+++ b/src/pybind/mgr/dashboard/frontend/cypress/integration/cluster/hosts.po.ts
@@ -127,8 +127,8 @@ export class HostsPageHelper extends PageHelper {
   @PageHelper.restrictTo(pages.index.url)
   maintenance(hostname: string, exit = false, force = false) {
     this.clearTableSearchInput();
+    this.getTableCell(this.columnIndex.hostname, hostname).click();
     if (force) {
-      this.getTableCell(this.columnIndex.hostname, hostname).click();
       this.clickActionButton('enter-maintenance');

       cy.get('cd-modal').within(() => {
@@ -145,7 +145,6 @@ export class HostsPageHelper extends PageHelper {
     }
     if (exit) {
       this.getTableCell(this.columnIndex.hostname, hostname)
-        .click()
         .parent()
         .find(`datatable-body-cell:nth-child(${this.columnIndex.status})`)
         .then(($ele) => {
@@ -163,7 +162,6 @@ export class HostsPageHelper extends PageHelper {
           expect(status).to.not.include('maintenance');
         });
     } else {
-      this.getTableCell(this.columnIndex.hostname, hostname).click();
       this.clickActionButton('enter-maintenance');

       this.getTableCell(this.columnIndex.hostname, hostname)
@@ -187,5 +185,9 @@ export class HostsPageHelper extends PageHelper {
       cy.wait(20000);
       this.expectTableCount('total', 0);
     });
+
+    // unselect it to avoid colliding with any other selection
+    // in different steps
+    this.getTableCell(this.columnIndex.hostname, hostname).click();
   }
 }
 ```

**Comment by Sarthak0702** [comment]
NA

**Comment by Sarthak0702** [comment]
N/A

**Comment by Sarthak0702** [comment]
jenkins test dashboard cephadm

**Comment by Sarthak0702** [comment]
jenkins test dashboard

**Comment by Sarthak0702** [comment]
jenkins test dashboard cephadm

**Comment by Sarthak0702** [comment]
jenkins test dashboard

**Comment by Sarthak0702** [comment]
jenkins test dashboard

**Comment by nizamial09** [comment]
```
diff --git a/src/pybind/mgr/dashboard/frontend/cypress/integration/cluster/services.po.ts b/src/pybind/mgr/dashboard/frontend/cypress/integration/cluster/services.po.ts
index cbc900813dc..b90c0ea5f07 100644
--- a/src/pybind/mgr/dashboard/frontend/cypress/integration/cluster/services.po.ts
+++ b/src/pybind/mgr/dashboard/frontend/cypress/integration/cluster/services.po.ts
@@ -166,6 +166,10 @@ export class ServicesPageHelper extends PageHelper {
     cy.get('cd-service-daemon-list').within(() => {
       this.getTableRow(daemon).click();
       this.clickActionButton(action);
+
+      // unselect it to avoid colliding with any other selection
+      // in different steps
+      this.getTableRow(daemon).click();
     });
   }
 }
 ```

**Comment by Sarthak0702** [comment]
jenkins test dashboard cephadm

**Comment by Sarthak0702** [comment]
jenkins test dashboard

**Comment by nizamial09** [comment]
jenkins test dashboard cephadm

**Comment by Sarthak0702** [comment]
jenkins test dashboard cephadm

**Comment by Sarthak0702** [comment]
jenkins test dashboard

**Comment by epuertat** [comment]
jenkins test dashboard

**Comment by epuertat** [comment]
jenkins test dashboard

**Comment by avanthakkar** [comment]
jenkins test dashboard

**Review by avanthakkar — APPROVED** [review]

**Review by nizamial09 — APPROVED** [review]

**Comment by epuertat** [comment]
jenkins test dashboard

**Comment by pereman2** [comment]
jenkins test dashboard

**Comment by epuertat** [comment]
jenkins test dashboard cephadm

**Comment by pereman2** [comment]
jenkins test dashboard cephadm

**Comment by pereman2** [comment]
jenkins test dashboard

**Comment by Sarthak0702** [comment]
jenkins test dashboard

**Comment by Sarthak0702** [comment]
jenkins test make check

**Comment by Sarthak0702** [comment]
jenkins test make check

**Comment by Sarthak0702** [comment]
jenkins test dashboard

**Comment by Sarthak0702** [comment]
jenkins test dashboard cephadm

**Comment by Sarthak0702** [comment]
jenkins test dashboard cephadm

**Comment by epuertat** [comment]
jenkins test api

**Comment by epuertat** [comment]
jenkins test dashboard cephadm

**Pull request merged**